src/Microsoft.ApplicationInsights/Managed/Shared/Extensibility/Implementation/Tracing/HeartbeatProvider.cs

```diff
@@ -0,0 +1,312 @@
+                        }
+
+                        return property;
+                    });
+
+                    return setResult;
+                }
+                catch (Exception e)
+                {
+                    CoreEventSource.Log.FailedToSetHeartbeatProperty(healthPropertyName, healthPropertyValue, isHealthy.HasValue ? isHealthy.Value.ToString() : "null", e.ToString());
+                }
+            }
+            else
+            {
+                CoreEventSource.Log.FailedToSetHeartbeatProperty(healthPropertyName, healthPropertyValue, isHealthy.HasValue ? isHealthy.Value.ToString() : "null");
+            }
+
+            return false;
+        }
+
+        #region IDisposable Support
+
+        public void Dispose()
+        {
+            this.Dispose(true);
+        }
+
+        /// <summary>
+        /// Get the metric telemetry item that will be sent.
+        /// 
+        /// Note: exposed to internal to allow inspection for testing.
+        /// </summary>
+        /// <returns>A MetricTelemtry item that contains the currently defined payload for a heartbeat 'pulse'.</returns>
+        internal ITelemetry GatherData()
+        {
+            var hbeat = new MetricTelemetry(heartbeatSyntheticMetricName, 0.0);
+
+            foreach (var payloadItem in this.heartbeatProperties)
+            {
+                hbeat.Properties.Add(payloadItem.Key, payloadItem.Value.PayloadValue);
+                hbeat.Sum += payloadItem.Value.IsHealthy ? 0 : 1;
+            }
+
+            hbeat.Sequence = string.Format(CultureInfo.CurrentCulture, "{0}", this.heartbeatsSent++);
+
+            return hbeat;
+        }
+
+        protected void Dispose(bool disposing)
+        {
+            if (!this.disposedValue)
+            {
+                if (disposing)
+                {
+                    if (this.HeartbeatTimer != null)
+                    {
+                        this.isEnabled = false;
+
+                        try
+                        {
+                            this.HeartbeatTimer.Dispose();
+                        }
+                        catch (Exception)
+                        {
+                            CoreEventSource.Log.LogError("Error occured when disposing heartbeat timer within HeartbeatProvider");
+                        }
+                    }
+                }
+
+                this.disposedValue = true;
+            }
+        }
+
+        #endregion
+
+        private void Send()
+        {
+            if (this.telemetryClient.TelemetryConfiguration.TelemetryChannel == null)
+            {
+                return;
+            }
+
+            var eventData = (MetricTelemetry)this.GatherData();
+
+            eventData.Context.Operation.SyntheticSource = heartbeatSyntheticMetricName;
+
+            this.telemetryClient.TrackMetric(eventData);
+        }
+
+        private void HeartbeatPulse(object state)
+        {
+            if (state is HeartbeatProvider)
```

Comment:
Can this entire method be wrapped inside a if (this.IsHeartbeatEnabled) {} ?

Reply:
Good call, simpler.